src/lib/crypto/sol-message-digest-impl-linux-kcapi.c

```diff
@@ -0,0 +1,957 @@
+error_accept:
+error_info:
+    errno_bkp = errno;
+    free(handle);
+    errno = errno_bkp;
+
+error_bfd:
+    errno_bkp = errno;
+    close(bfd);
+    errno = errno_bkp;
+    return NULL;
+}
+
+static void
+_sol_message_digest_free(struct sol_message_digest *handle)
+{
+    struct sol_message_digest_pending_feed *pf;
+    uint16_t i;
+
+    SOL_DBG("free handle %p fd=%d, pending_feed=%hu, digest=%p",
+        handle, handle->fd, handle->pending_feed.len, handle->digest);
+
+    _sol_message_digest_thread_fini(handle);
+
+    SOL_VECTOR_FOREACH_IDX (&handle->pending_feed, pf, i) {
+        sol_blob_unref(pf->blob);
+    }
+    sol_vector_clear(&handle->pending_feed);
+
+    if (handle->digest)
+        sol_blob_unref(handle->digest);
+
+    close(handle->fd);
+    free(handle);
+}
+
+static inline void
+_sol_message_digest_unref(struct sol_message_digest *handle)
+{
+    handle->refcnt--;
+    if (handle->refcnt == 0)
+        _sol_message_digest_free(handle);
+}
+
+static inline void
+_sol_message_digest_ref(struct sol_message_digest *handle)
+{
+    handle->refcnt++;
+}
+
+SOL_API void
+sol_message_digest_del(struct sol_message_digest *handle)
+{
+    SOL_NULL_CHECK(handle);
+    SOL_EXP_CHECK(handle->deleted);
+    SOL_INT_CHECK(handle->refcnt, < 1);
+
+    handle->deleted = true;
+
+    _sol_message_digest_thread_stop(handle);
+
+    SOL_DBG("del handle %p fd=%d, refcnt=%" PRIu32
+        ", pending_feed=%hu, digest=%p",
+        handle, handle->fd, handle->refcnt,
+        handle->pending_feed.len, handle->digest);
+    _sol_message_digest_unref(handle);
+}
+
+static void
+_sol_message_digest_setup_receive_digest(struct sol_message_digest *handle)
+{
+    void *mem;
+
+    if (handle->digest) {
+        SOL_WRN("handle %p fd=%d already have a digest to be received (%p).",
+            handle, handle->fd, handle->digest);
+        return;
+    }
+
+    mem = malloc(handle->digest_size);
+    SOL_NULL_CHECK(mem);
+
+    handle->digest = sol_blob_new(SOL_BLOB_TYPE_DEFAULT, NULL,
+        mem, handle->digest_size);
+    SOL_NULL_CHECK_GOTO(handle->digest, error);
+
+    handle->digest_offset = 0;
+
+    SOL_DBG("handle %p fd=%d to receive digest of %zd bytes at blob %p mem=%p",
+        handle, handle->fd, handle->digest_size,
+        handle->digest, handle->digest->mem);
+
+    return;
+
+error:
+    free(mem);
+}
+
+static void
+_sol_message_digest_report_feed_blob(struct sol_message_digest *handle, struct sol_blob *input)
+{
+#ifdef MESSAGE_DIGEST_USE_THREAD
+    struct sol_message_digest_pending_dispatch *pd;
+
+    _sol_message_digest_lock(handle);
+
+    pd = sol_vector_append(&handle->pending_dispatch);
+    SOL_NULL_CHECK_GOTO(pd, error);
+    pd->blob = input;
+    pd->is_digest = false;
+
+    _sol_message_digest_unlock(handle);
+    sol_worker_thread_feedback(handle->thread);
+    return;
+
+error:
+    _sol_message_digest_unlock(handle);
+    sol_blob_unref(input); /* this may cause problems if main thread changes blob refcnt */
```

Comment:
it won't be the correct thread, we're from a worker thread trying to unref the input blob... this should happen from the main thread, but we can't allocate memory to append this request for the main thread -- it's a cumbersome situation.

The other way is to leak the input blob.